File: Tests/Review2/bfs-dfs/bfs-dfs.py
from typing import List


class TestNode:

    def __init__(self, value: int = 0, neighbors: List = []):
        self.value: int = value
        self.neighbors: List[TestNode] = neighbors


def BFS(node: TestNode):
    if not node:
        return

    queue = [node]
    layer = 0

    while len(queue):
        size = len(queue)
        for _ in range(0, size):
            current = queue.pop(0)
            print(current.value)
            for neighbor in current.neighbors:
                queue.append(neighbor)
            layer += 1


def DFS_Stack(node: TestNode):
    if not node:
        return

    stack = [node]

    while len(stack):
        current = stack.pop()
        print(current.value)
        if len(node.neighbors) > 0:
            for child in current.neighbors:
                stack.append(child)


def DFS_Recursive(node: TestNode):

    def recursive(node: TestNode):
        if not node:
            return

        print(node.value)
        if len(node.neighbors) > 0:
            for child in node.neighbors:
                recursive(child)

    recursive(node)

File: Tests/Review2/bfs-dfs/bfs-dfs.ts
interface TestNode {
  value: any;
  neighbors: TestNode[];
}

export function BFS(root: TestNode): number {
  if (!root) {
    return 0;
  }

  const queue: TestNode[] = [root];
  let layer = 0;

  while (queue.length) {
    const size = queue.length;
    for (let _ = 0; _ < size; _++) {
      const current = queue.shift()!;
      console.log(current.value);

      for (const neighbor of current.neighbors) {
        queue.push(neighbor);
      }
    }
    layer++;
  }

  return layer;
}

export function DFS_Iteration(root: TestNode) {
  if (!root) {
    return;
  }

  const stack: TestNode[] = [root];

  while (stack.length) {
    const current = stack.pop()!;
    console.log(current.value);

    for (const neighbor of current.neighbors) {
      stack.push(neighbor);
    }
  }
}

export function DFS_Recursion(root: TestNode) {
  if (!root) {
    return;
  }

  const recursive = (node: TestNode) => {
    if (!node) {
      return;
    }
    console.log(node.value);

    for (const neighbor of node.neighbors) {
      recursive(neighbor);
    }
  };

  recursive(root);
}

// test
const root: TestNode = {
  value: 1,
  neighbors: [
    {
      value: 2,
      neighbors: [
        {
          value: 4,
          neighbors: [],
        },
        {
          value: 5,
          neighbors: [],
        },
      ],
    },
    {
      value: 3,
      neighbors: [
        {
          value: 6,
          neighbors: [
            {
              value: 7,
              neighbors: [],
            },
          ],
        },
      ],
    },
  ],
};

// BFS
const res = BFS(root);
console.log('layer', res);

// DFS Recursion
console.log('------');
DFS_Recursion(root);

// DFS Iteration
console.log('------');
DFS_Iteration(root);
